starknet_py/class_hash.py

```diff
@@ -0,0 +1,124 @@
+import json
+from typing import List
+
+from starkware.cairo.common.hash_state import compute_hash_on_elements
+from starkware.starknet.public.abi import starknet_keccak
+
+from starknet_py.cairo.felt import encode_shortstring
+from starknet_py.constants import API_VERSION
+from starknet_py.net.client_models import DeclaredContract, EntryPoint
+
+
+def compute_class_hash(contract_class: DeclaredContract) -> int:
+    """
+    Calculate class hash of a declared contract.
+    """
+    api_version = API_VERSION
+
+    _entry_points = contract_class.entry_points_by_type
+
+    external_entry_points_hash = compute_hash_on_elements(
+        _entry_points_array(_entry_points.external)
+    )
+    l1_handler_entry_points_hash = compute_hash_on_elements(
+        _entry_points_array(_entry_points.l1_handler)
+    )
+    constructor_entry_points_hash = compute_hash_on_elements(
+        _entry_points_array(_entry_points.constructor)
+    )
+
+    _encoded_builtins = [
+        encode_shortstring(builtin) for builtin in contract_class.program["builtins"]
+    ]
+    builtins_hash = compute_hash_on_elements(_encoded_builtins)
+
+    hinted_class_hash = _compute_hinted_class_hash(contract_class)
+
+    program_data_hash = compute_hash_on_elements(
+        [int(data_, 0) for data_ in contract_class.program["data"]]
+    )
+
+    return compute_hash_on_elements(
+        [
+            api_version,
+            external_entry_points_hash,
+            l1_handler_entry_points_hash,
+            constructor_entry_points_hash,
+            builtins_hash,
+            hinted_class_hash,
+            program_data_hash,
+        ]
+    )
+
+
+def _entry_points_array(entry_points: List[EntryPoint]) -> List[int]:
+    entry_points_array = []
+    for entry_point in entry_points:
+        entry_points_array.extend([entry_point.selector, entry_point.offset])
+
+    return entry_points_array
+
+
+def _compute_hinted_class_hash(contract_class: DeclaredContract) -> int:
+    program = contract_class.program
+    program["debug_info"] = None
+
+    if "attributes" in program:
+        program = _delete_attributes(program)
+
+    # If compiler_version is not present, this was compiled with a compiler before version 0.10.0.
+    # Use "(a : felt)" syntax instead of "(a: felt)" so that the class hash will be the same.
+    if "compiler_version" not in program:
+        program["identifiers"] = _fix_cairo_types(program["identifiers"])
+
+    # ContractClass stores keys of `hints` field as integers, resulting in different sorting order
+    program["hints"] = {int(key): val for key, val in program["hints"].items()}
+
+    class_ = dict(program=program, abi=contract_class.abi)
+    serialized_contract_class = json.dumps(obj=class_, sort_keys=True)
```

Comment:
Let's verify if we really need to sort the keys.

Reply:
Turned out that the output from `starknet-compile` has all keys sorted and the only keys not sorted at this point in code were `program` and `abi` (dict with them is created in the line before).